ns/nslogic.go

```diff
@@ -101,21 +104,70 @@ type PgoPgRoleBinding struct {
 	TargetNamespace string
 }
 
-// CreateNamespace ...
-func CreateNamespace(clientset *kubernetes.Clientset, installationName, pgoNamespace, createdBy, newNs string) error {
+// NamespaceOperatingMode defines the different namespace operating modes for the Operator
+type NamespaceOperatingMode string
+
+const (
+	// NamespaceOperatingModeDynamic enables full dynamic namespace capabilities, in which the
+	// Operator can create, delete and update any namespaces within the Kubernetes cluster, while
+	// then also having the ability to create the roles, role bindings and service accounts within
+	// those namespaces as required for the Operator to create PG clusters.  Additionally, while in
+	// this mode the Operator can listen for namespace events (e.g. namespace additions, updates
+	// and deletions), and then create or remove controllers for various namespaces as those
+	// namespaces are added or removed from the Kubernetes cluster.
+	NamespaceOperatingModeDynamic NamespaceOperatingMode = "dynamic"
+	// NamespaceOperatingModeReadOnly allows the Operator to listen for namespace events within the
+	// Kubernetetes cluster, and then create and run and/or remove controllers as namespaces are
+	// added and deleted.  However, while in this mode the Operator is unable to create, delete or
+	// update namespaces, nor can it create the RBAC it requires in any of those namespaces to
+	// create PG clusters.  Therefore,  while in a "readonly" mode namespaces must be
+	// pre-configured with the proper RBAC, since the Operator cannot create the RBAC itself.
+	NamespaceOperatingModeReadOnly NamespaceOperatingMode = "readonly"
+	// NamespaceOperatingModeDisabled causes namespace capabilities to be disabled altogether.  In
+	// this mode the Operator will simply attempt to work with the target namespaces specified
+	// during installation.  If no target namespaces are specified, then it will be configured to
+	// work within the namespace in which the Operator is deployed.
+	NamespaceOperatingModeDisabled NamespaceOperatingMode = "disabled"
+
+	// DNS-1123 formatting and error message for validating namespace names
+	dns1123Fmt    string = "[a-z0-9]([-a-z0-9]*[a-z0-9])?"
+	dns1123ErrMsg string = "A namespace name must consist of lower case" +
+		"alphanumeric characters or '-', and must start and end with an alphanumeric character"
```

Comment:
I would recommend using the standard Kubernetes validation library:

- [https://godoc.org/k8s.io/kubernetes/pkg/apis/core/validation](https://godoc.org/k8s.io/kubernetes/pkg/apis/core/validation)
- [https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go)

Reply:
Per discussion, ignore.